ethdb/pebble/pebble.go

```diff
@@ -0,0 +1,571 @@
+			compWrite += int64(levelMetrics.BytesCompacted)
+			compRead += int64(levelMetrics.BytesRead)
+		}
+
+		nWrite += int64(metrics.WAL.BytesWritten)
+
+		compWrites[i%2] = compWrite
+		compReads[i%2] = compRead
+		nWrites[i%2] = nWrite
+
+		if d.writeDelayNMeter != nil {
+			d.writeDelayNMeter.Mark(writeDelayCounts[i%2] - writeDelayCounts[(i-1)%2])
+		}
+		if d.writeDelayMeter != nil {
+			d.writeDelayMeter.Mark(writeDelayTimes[i%2] - writeDelayTimes[(i-1)%2])
+		}
+		if d.compTimeMeter != nil {
+			d.compTimeMeter.Mark(compTimes[i%2] - compTimes[(i-1)%2])
+		}
+		if d.compReadMeter != nil {
+			d.compReadMeter.Mark(compReads[i%2] - compReads[(i-1)%2])
+		}
+		if d.compWriteMeter != nil {
+			d.compWriteMeter.Mark(compWrites[i%2] - compWrites[(i-1)%2])
+		}
+		if d.diskSizeGauge != nil {
+			d.diskSizeGauge.Update(int64(metrics.DiskSpaceUsage()))
+		}
+		if d.diskReadMeter != nil {
+			d.diskReadMeter.Mark(0) // pebble doesn't track non-compaction reads
+		}
+		if d.diskWriteMeter != nil {
+			d.diskWriteMeter.Mark(nWrites[i%2] - nWrites[(i-1)%2])
+		}
+		// See https://github.com/cockroachdb/pebble/pull/1628#pullrequestreview-1026664054
+		manuallyAllocated := metrics.BlockCache.Size + int64(metrics.MemTable.Size) + int64(metrics.MemTable.ZombieSize)
+		d.manualMemAllocGauge.Update(manuallyAllocated)
+		d.memCompGauge.Update(metrics.Flush.Count)
+		d.nonlevel0CompGauge.Update(nonLevel0CompCount)
+		d.level0CompGauge.Update(level0CompCount)
+		d.seekCompGauge.Update(metrics.Compact.ReadCount)
+
+		// Sleep a bit, then repeat the stats collection
+		select {
+		case errc = <-d.quitChan:
+			// Quit requesting, stop hammering the database
+		case <-timer.C:
+			timer.Reset(refresh)
+			// Timeout, gather a new set of stats
+		}
+	}
+	errc <- nil
+}
+
+// batch is a write-only batch that commits changes to its host database
+// when Write is called. A batch cannot be used concurrently.
+type batch struct {
+	b    *pebble.Batch
+	size int
+}
+
+// Put inserts the given value into the batch for later committing.
+func (b *batch) Put(key, value []byte) error {
+	b.b.Set(key, value, nil)
+	b.size += len(key) + len(value)
+	return nil
+}
+
+// Delete inserts the a key removal into the batch for later committing.
+func (b *batch) Delete(key []byte) error {
+	b.b.Delete(key, nil)
+	b.size += len(key)
+	return nil
+}
+
+// ValueSize retrieves the amount of data queued up for writing.
+func (b *batch) ValueSize() int {
+	return b.size
+}
+
+// Write flushes any accumulated data to disk.
+func (b *batch) Write() error {
+	return b.b.Commit(pebble.NoSync)
+}
+
+// Reset resets the batch for reuse.
+func (b *batch) Reset() {
+	b.b.Reset()
+	b.size = 0
+}
+
+// Replay replays the batch contents.
+func (b *batch) Replay(w ethdb.KeyValueWriter) error {
+	reader := b.b.Reader()
+	for {
+		kind, k, v, ok := reader.Next()
+		if !ok {
+			break
+		}
+		// I have no idea whether the iterated key and value
+		// are safe to use, deep copy them temporarily.
```

Comment:
https://github.com/cockroachdb/pebble/blob/c54158bff06a/batch.go#L180

> // The internal batch representation is the on disk format for a batch in the
// WAL, and thus stable. New record kinds may be added, but the existing ones
// will not be modified.

Would seem to me that the copy is not needed?

Reply:
Are we certain that we, in our `w.Put(...)`-method do not mutate the slice? And if we do, are we allowed to?

Reply:
We don't mutate it, but we stuff it into a cache:
```
		c.db.cleans.Set(hash[:], rlp)
```
Might be fine to do as you propose